File: MarsRover/MarsPhoto.m
//
//  MarsPhoto.m
//  MarsRover
//
//  Created by Carolyn Lea on 10/10/18.
//  Copyright © 2018 Carolyn Lea. All rights reserved.
//

#import "MarsPhoto.h"

@implementation MarsPhoto

- (instancetype)initWithIdentifier:(NSNumber *)identifier
                               sol:(NSNumber *)sol
                            camera:(Camera *)camera
                         earthDate:(NSDate *)earthDate
                          imageURL:(NSURL *)imageURL
{
    self = [super init];
    if (self)
    {
        
    }
    return self;
}
- (instancetype)initWithDictionary:(NSDictionary<NSString *,id> *)dictionary
{
    NSNumber *identifier = dictionary[@"id"];
    NSNumber *sol = dictionary[@"sol"];
    Camera *camera = dictionary[@"camera"];
    NSDate *earthDate = dictionary[@"earth_date"];
    NSURL *imageURL = dictionary[@"img_src"];
    
    if (!identifier || !sol || !camera || !earthDate || !imageURL)
    {
        return nil;
    }
    
    return [self initWithIdentifier:identifier sol:sol camera:camera earthDate:earthDate imageURL:imageURL];
}

@end

File: MarsRover/MarsPhoto.swift
//
//  MarsPhoto.swift
//  MarsRover
//
//  Created by Carolyn Lea on 10/9/18.
//  Copyright © 2018 Carolyn Lea. All rights reserved.
//

import Foundation

struct MarsPhotoReference: Codable
{
    let id: Int
    let sol: Int
    let camera: Camera
    let earthDate: Date
    
    let imageURL: URL
    
    enum CodingKeys: String, CodingKey
    {
        case id
        case sol
        case camera
        case earthDate
        case imageURL = "imgSrc"
    }
    
    static let dateFormatter: DateFormatter =
    {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.timeZone = TimeZone(abbreviation: "GMT")
        formatter.dateFormat = "yyyy-MM-dd"
        return formatter
    }()
    
    static var jsonDecoder: JSONDecoder
    {
        let result = JSONDecoder()
        result.keyDecodingStrategy = .convertFromSnakeCase
        result.dateDecodingStrategy = .formatted(dateFormatter)
        return result
    }
}